Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XldbProtobufExtensions.cs

```diff
@@ -558,5 +548,479 @@ public static Xldb.Fingerprint ToFingerprint(this Fingerprint fingerprint)
+        {
+            var xldbModulePip = new Xldb.ModulePip
+            {
+                ParentPipInfo = parentPip,
+                Module = pip.Module.Value.Value,
+                Identity = pip.Identity.ToString(),
+                ResolverKind = pip.ResolverKind.ToString(),
+                ResolverName = pip.ResolverName.ToString(),
+                Version = pip.Version.ToString(),
+                Location = new Xldb.LocationData()
+                {
+                    IsValid = pip.Location.IsValid,
+                    Line = pip.Location.Line,
+                    Path = pip.Location.Path.ToAbsolutePath(pathTable),
+                    Position = pip.Location.Position
+                },
+                Provenance = pip.Provenance.ToPipProvenance(),
+                PipType = (PipType)pip.PipType
+            };
+
+            if (pip.Tags.IsValid)
+            {
+                xldbModulePip.Tags.AddRange(pip.Tags.Select(key => key.ToString()));
+            }
+
+            return xldbModulePip;
+        }
+
+        /// <nodoc />
+        public static Xldb.SealDirectory ToSealDirectory(this SealDirectory pip, PathTable pathTable, Xldb.Pip parentPip)
+        {
+            var xldbSealDirectory = new Xldb.SealDirectory
+            {
+                ParentPipInfo = parentPip,
+                Kind = (SealDirectoryKind)pip.Kind,
+                DirectoryRoot = pip.DirectoryRoot.ToAbsolutePath(pathTable),
+                IsComposite = pip.IsComposite,
+                Scrub = pip.Scrub,
+                IsInitialzed = pip.IsInitialized,
+                Directory = pip.Directory.ToDirectoryArtifact(pathTable),
+                IsSealSourceDirectory = pip.IsSealSourceDirectory,
+                Provenance = pip.Provenance.ToPipProvenance(),
+                PipType = (PipType)pip.PipType
+
+            };
+
+            xldbSealDirectory.Patterns.AddRange(pip.Patterns.Select(key => key.ToString()));
+            xldbSealDirectory.Contents.AddRange(pip.Contents.Select(file => file.ToFileArtifact(pathTable)));
+            xldbSealDirectory.ComposedDirectories.AddRange(pip.ComposedDirectories.Select(dir => dir.ToDirectoryArtifact(pathTable)));
+
+            if (pip.Tags.IsValid)
+            {
+                xldbSealDirectory.Tags.AddRange(pip.Tags.Select(key => key.ToString()));
+            }
+
+            return xldbSealDirectory;
+        }
+
+        /// <nodoc />
+        public static Xldb.CopyFile ToCopyFile(this CopyFile pip, PathTable pathTable, Xldb.Pip parentPip)
+        {
+            var xldbCopyFile = new Xldb.CopyFile
+            {
+                ParentPipInfo = parentPip,
+                Source = pip.Source.ToFileArtifact(pathTable),
+                Destination = pip.Destination.ToFileArtifact(pathTable),
+                OutputsMustRemainWritable = pip.OutputsMustRemainWritable,
+                Provenance = pip.Provenance.ToPipProvenance(),
+                PipType = (PipType)pip.PipType

```

Comment:
>PipType [](start = 16, length = 7)

Isn't this redundant based on the type?

Reply:
Ordinarily this would be an overriding attribute due to inheritance. I'm going to go ahead and just remove it from the "parent pip" object since we don't need it in there. 

---
In reply to: [314447532](https://github.com/microsoft/BuildXL/pull/724#discussion_r314447532) [](ancestors = 314447532)